File: backend/services/compilerService.js
import axios from 'axios';

// Switch to public free instance
const JUDGE0_BASE_URL = process.env.JUDGE0_BASE_URL || 'https://ce.judge0.com';

const LANGUAGE_MAP = {
  javascript: 63,
  python: 71, // Python 3
  cpp: 54,    // C++ (GCC 9.2.0)
};

const getHeaders = () => {
  const headers = {
    'Content-Type': 'application/json',
  };
  
  const apiKey = process.env.JUDGE0_API_KEY;
  if (apiKey) {
    if (JUDGE0_BASE_URL.includes('rapidapi.com')) {
      headers['X-RapidAPI-Key'] = apiKey;
      headers['X-RapidAPI-Host'] = new URL(JUDGE0_BASE_URL).hostname;
    } else {
      headers['X-Auth-Token'] = apiKey;
    }
  }
  return headers;
};

/**
 * Submits code to Judge0 and returns a token immediately (Asynchronous).
 * This leverages Judge0's internal Redis queue.
 */
export const submitToJudge = async (code, language, input, expectedOutput = null) => {
  const languageId = LANGUAGE_MAP[language.toLowerCase()];
  if (!languageId) throw new Error(`Unsupported language: ${language}`);

  try {
    const response = await axios.post(
      `${JUDGE0_BASE_URL}/submissions?base64_encoded=true&wait=false`,
      {
        source_code: Buffer.from(code).toString('base64'),
        language_id: languageId,
        stdin: Buffer.from(input).toString('base64'),
        expected_output: expectedOutput ? Buffer.from(expectedOutput).toString('base64') : null,
        cpu_time_limit: 2.0, // 2 seconds limit
        memory_limit: 128000, // 128MB limit
      },
      {
        headers: getHeaders(),
        timeout: 10000,
      }
    );

    return response.data.token;
  } catch (error) {
    const status = error.response?.status;
    console.error('Judge0 submission error:', error.response?.data || error.message);
    throw new Error(status === 429 ? 'Service busy, please try again.' : 'Code execution service unavailable.');
  }
};

/**
 * Retrieves the result of a submission by its token.
 */
export const getJudgeResult = async (token) => {
  try {
    const response = await axios.get(
      `${JUDGE0_BASE_URL}/submissions/${token}?base64_encoded=true`,
      {
        headers: getHeaders(),
        timeout: 5000,
      }
    );

    const { status, stdout, stderr, compile_output, time, memory } = response.data;

    // Status codes: 1=Queue, 2=Processing. 
    // If it's still being processed, return the status without decoding.
    if (status.id <= 2) {
      return { 
        isProcessing: true,
        status: status.description,
        token 
      };
    }

    // Process finished: Accept, WA, TLE, etc.
    return {
      isProcessing: false,
      status: status.description,
      statusId: status.id,
      stdout: stdout ? Buffer.from(stdout, 'base64').toString() : '',
      stderr: stderr ? Buffer.from(stderr, 'base64').toString() : '',
      compile_output: compile_output ? Buffer.from(compile_output, 'base64').toString() : '',
      time,
      memory,
    };
  } catch (error) {
    console.error('Judge0 retrieval error:', error.response?.data || error.message);
    throw new Error('Failed to retrieve execution results.');
  }
};

/**
 * BACKWARD COMPATIBILITY: Executes code synchronously using wait=true.
 * Still useful for non-interactive scripts or small runs.
 */
export const executeCode = async (code, language, input, expectedOutput = null, retries = 2) => {
  const languageId = LANGUAGE_MAP[language.toLowerCase()];
  if (!languageId) throw new Error(`Unsupported language: ${language}`);

  for (let attempt = 1; attempt <= retries + 1; attempt++) {
    try {
      const response = await axios.post(
        `${JUDGE0_BASE_URL}/submissions?base64_encoded=true&wait=true`,
        {
          source_code: Buffer.from(code).toString('base64'),
          language_id: languageId,
          stdin: Buffer.from(input).toString('base64'),
          expected_output: expectedOutput ? Buffer.from(expectedOutput).toString('base64') : null,
          cpu_time_limit: 2.0,
          memory_limit: 128000,
        },
        {
          headers: getHeaders(),
          timeout: 10000,
        }
      );

      const { status, stdout, stderr, compile_output, time, memory } = response.data;

      return {
        status: status.description,
        stdout: stdout ? Buffer.from(stdout, 'base64').toString() : '',
        stderr: stderr ? Buffer.from(stderr, 'base64').toString() : '',
        compile_output: compile_output ? Buffer.from(compile_output, 'base64').toString() : '',
        time,
        memory,
      };
    } catch (error) {
      const status = error.response?.status;
      if (status === 429 && attempt <= retries) {
        await new Promise(resolve => setTimeout(resolve, 2000 * attempt));
        continue;
      }
      if (attempt <= retries) {
        await new Promise(resolve => setTimeout(resolve, 1000));
        continue;
      }
      throw new Error(status === 429 ? 'Service busy, please try again.' : 'Code execution service unavailable.');
    }
  }
};

File: backend_java/src/main/java/com/learnsphere/service/CompilerService.java
package com.learnsphere.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

@Service
public class CompilerService {

    @Value("${JUDGE0_BASE_URL:https://ce.judge0.com}")
    private String judge0BaseUrl;

    @Value("${JUDGE0_API_KEY:}")
    private String judge0ApiKey;

    private final RestTemplate restTemplate = new RestTemplate();

    private static final Map<String, Integer> LANGUAGE_MAP = new HashMap<>();
    static {
        LANGUAGE_MAP.put("javascript", 63);
        LANGUAGE_MAP.put("python", 71);
        LANGUAGE_MAP.put("cpp", 54);
    }

    private HttpHeaders getHeaders() {
        HttpHeaders headers = new HttpHeaders();
        headers.set("Content-Type", "application/json");

        if (judge0ApiKey != null && !judge0ApiKey.trim().isEmpty()) {
            if (judge0BaseUrl.contains("rapidapi.com")) {
                headers.set("X-RapidAPI-Key", judge0ApiKey);
                try {
                    String host = new java.net.URL(judge0BaseUrl).getHost();
                    headers.set("X-RapidAPI-Host", host);
                } catch (Exception ignored) {}
            } else {
                headers.set("X-Auth-Token", judge0ApiKey);
            }
        }
        return headers;
    }

    private String encode(String text) {
        if (text == null) return null;
        return Base64.getEncoder().encodeToString(text.getBytes());
    }

    private String decode(String base64) {
        if (base64 == null || base64.isEmpty()) return "";
        try {
            return new String(Base64.getDecoder().decode(base64));
        } catch (Exception e) {
            return base64; // Return as-is if decoding fails
        }
    }

    public String submitToJudge(String code, String language, String input) {
        return submitToJudge(code, language, input, null);
    }

    public String submitToJudge(String code, String language, String input, String expectedOutput) {
        Integer languageId = LANGUAGE_MAP.get(language.toLowerCase());
        if (languageId == null) throw new RuntimeException("Unsupported language: " + language);

        Map<String, Object> body = new HashMap<>();
        body.put("source_code", encode(code));
        body.put("language_id", languageId);
        body.put("stdin", encode(input));
        if (expectedOutput != null) {
            body.put("expected_output", encode(expectedOutput));
        }
        body.put("cpu_time_limit", 2.0);
        body.put("memory_limit", 128000);

        HttpEntity<Map<String, Object>> entity = new HttpEntity<>(body, getHeaders());

        try {
            String url = judge0BaseUrl + "/submissions?base64_encoded=true&wait=false";
            ResponseEntity<Map> response = restTemplate.postForEntity(url, entity, Map.class);
            return (String) response.getBody().get("token");
        } catch (Exception e) {
            System.err.println("Judge0 submission error: " + e.getMessage());
            throw new RuntimeException("Code execution service unavailable.");
        }
    }

    public Map<String, Object> getJudgeResult(String token) {
        try {
            String url = judge0BaseUrl + "/submissions/" + token + "?base64_encoded=true";
            HttpEntity<String> entity = new HttpEntity<>(getHeaders());
            
            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, entity, Map.class);
            Map<String, Object> data = response.getBody();

            Map<String, Object> statusObj = (Map<String, Object>) data.get("status");
            int statusId = (Integer) statusObj.get("id");
            String description = (String) statusObj.get("description");

            Map<String, Object> result = new HashMap<>();

            // Status codes: 1=Queue, 2=Processing
            if (statusId <= 2) {
                result.put("isProcessing", true);
                result.put("status", description);
                result.put("token", token);
                return result;
            }

            result.put("isProcessing", false);
            result.put("status", description);
            result.put("statusId", statusId);
            result.put("stdout", decode((String) data.get("stdout")));
            result.put("stderr", decode((String) data.get("stderr")));
            result.put("compile_output", decode((String) data.get("compile_output")));
            result.put("time", data.get("time"));
            result.put("memory", data.get("memory"));

            return result;
        } catch (Exception e) {
            System.err.println("Judge0 retrieval error: " + e.getMessage());
            throw new RuntimeException("Failed to retrieve execution results.");
        }
    }
}
